src/high-level/polynomial-solver.lisp

```diff
@@ -0,0 +1,112 @@
+
+           (make-companion-matrix (polynomial)
+             "Return the companion matrix of POLYNOMIAL."
+             (loop :with polynomial := (polynomial-normalize polynomial)
+                   :with coefficients := (polynomial-coefficients polynomial)
+                   :with n := (1- (length coefficients))
+                   :with matrix := (make-zero-matrix n n)
+                   :for i :below n
+                   :for coefficient :of-type (complex double-float) := (aref coefficients i)
+                   :when (plusp i) :do (setf (ref matrix i (1- i)) 1.0d0)
+                   :do (setf (ref matrix i (1- n)) (- coefficient))
+                   :finally (return matrix))))
+
+    (nth-value 0 (eig (make-companion-matrix polynomial)))))
+
+(defun polynomial-eval (polynomial value)
+  "Return the result of evaluating POLYNOMIAL at VALUE using Horner's rule."
+  (declare (type polynomial polynomial)
+           (type (complex double-float) value))
+
+  (let ((coefficients (polynomial-coefficients polynomial)))
+    (do* ((i 0 (1+ i))
+          (c (aref coefficients 0) (aref coefficients i))
+          (x (complex 1.0d0) (* x value))
+          (y (* x c) (+ y (* x c))))
+         ((= i (1- (length coefficients))) y)
+
+      (declare (type fixnum i)
+               (type (complex double-float) x y)))))
+
+(defun polynomial-diff (polynomial)
+  "Return the polynomial corresponding to the derivative of POLYNOMIAL."
+  (declare (type polynomial polynomial)
+           (values polynomial))
+
+  (loop :with coefficients := (polynomial-coefficients polynomial)
+        :with n := (length coefficients)
+        :with new-coefficients := (subseq coefficients 1 n)
+        :for i :below (1- n)
+        :for c := (aref new-coefficients i)
+        :do (setf (aref new-coefficients i) (* (1+ i) c))
+        :finally (return (%make-polynomial :coefficients new-coefficients))))
+
+(defconstant +default-tolerance+ (* 1e2 +double-float-epsilon+)
+  "Default tolerance for Newton's method.")
+
+(defconstant +default-max-iterations+ 100
+  "Default maximum number of iterations in Newton's method.")
+
+(defun polynomial-newton-iteration (polynomial initial-value
+                                    &key (tolerance +default-tolerance+)
+                                      (max-iterations +default-max-iterations+))
+  "Return the result of running Newton's method on POLYNOMIAL starting with INITIAL-VALUE. The arguments TOLERANCE and MAX-ITERATIONS determine the convergence criterion."
+  (declare (type polynomial polynomial)
+           (type (complex double-float) initial-value)
+           (type double-float tolerance)
+           (type fixnum max-iterations)
+           (values (complex double-float)))
+
+  (loop :with diff :of-type polynomial := (polynomial-diff polynomial)
+        :with x :of-type (complex double-float) := initial-value
+        :for i :of-type fixnum :below (min 1 max-iterations)
```

Comment:
confused. why `(min 1 ..)` here?

Reply:
oh, just realized that `i` is unused? maybe `:repeat` ?

Reply:
> Newton's method is the only thing I recognized here because SICP, but I trust that you and ecp have the polynomial stuff covered (plus tests cases). LGTM.

Thank you for checking it despite it being unfamiliar. I value your feedback.

Reply:
Good point. However, when the `loop` using `:repeat` macroexpands, it inserts a call to `ceiling` so I'd rather let it be.

Reply:
It turns out that when optimizing for speed during compilation, the type system elides the call to `ceiling` so I went with your suggestion of using `:repeat`. Thanks.